src/box/lua/console.c

```diff
@@ -314,6 +320,8 @@ console_hide_prompt(void)
 	if (!console_can_hide_show_prompt())
 		return;
 
+	tt_pthread_mutex_lock(&prompt_mutex);
```

Comment:
AFAICS readline doesn't use any mutexes internally, e.g. see [`rl_replace_line`](http://git.savannah.gnu.org/cgit/readline.git/tree/text.c#n212). Doesn't it mean that there may be a race between the tx thread calling an `rl_*` function while processing a console event and another thread calling another `rl_*` function from this callback?

Reply:
https://lists.gnu.org/archive/html/bug-readline/2006-11/msg00014.html

So, you suggest wrapping all work with readline with this mutex?

Reply:
I'd rather suggest reverting the original patch, because TBCH I don't quite understand the benefit of it. Logs should be written to a file/socket, not to stderr. As for Lua `print`, we should probably just redirect `print` output to a separate file/socket. Also, AFAIR there was an idea to override `print` to be handled properly by a console client, both local and remote.

Wrapping a call to every single `rl_*` function would look ugly IMO, besides we'd take and release the mutex multiple times per each console event. We could probably protect the whole `lbox_console_readline` with the mutex... Anyhow, this would be an anti-pattern, because we'd protect code sections with a mutex, not data. It's easy to overlook a code section, which should be protected. Even in this commit, we protect code instead of data, which doesn't look good. It'd probably be best if we followed the "Tarantool way" and, instead of using mutexes, sent an event to the tx thread.

Reply:
@Totktonada RFC

Reply:
Personally I meet the problem while playing with https://github.com/tarantool/crud/blob/master/doc/playground.lua (try to run it and type some commands). It is not a production environment, where we have logs redirected to syslog/journald/file, but a developer playground, where we want all the information present on a single terminal. It would be nice to don't mix prompt/stdout/stderr if it is simple to achieve. However:

* The fix is partial, because it doesn't handle the incremental search (Ctrl+R).
* Even this partial fix is not really simple, rather hacky.
* This code is now known to lead to problems (which you're fixing here) and it lets me suspect that there may be more hidden problems.
* It becomes even more complex with the fixes (disregarding of a way to fix).

In brief: I disagree that there are no benefits, but I agree that they're not so large comparing to risks and maintenance costs.

I would also ask @Mons as a person who initially highlighted the problem.

Reply:
IOW, I'm okay with attempt to fix it, but I'm also okay if we'll decide to don't support it.

(TBH, I'm wonder why terminal emulators have no API for a status line/user input/prompt and so on. It would eliminate a need to hack it around in readline; at least for some terminals -- other ones would follow it later. The output and prompt/input/status would be completely separate.)

Reply:
Discussed with @locker, @alyapunov, @Mons, @igormunkin, @Totktonada: decided that fixing this feature to work for multiple thread is not acceptable right now — this feature will only work in main (transaction) thread.